### PR TITLE
Support proactive checks in overload manager Api

### PR DESCRIPTION
**Pull request by nezdolik**
Signed-off-by: Kateryna Nezdolii <nezdolik@spotify.com>

This PR adds proacrive checks into overload manager/thread local state, when callers of overload manager API can check resource usage instantaneously on demand. This is first PR out of series (full code here: https://github.com/envoyproxy/envoy/pull/15707). next PR will include new proactive monitor for downstream connections. 

Commit Message: Support reactive checks in overload manager Api
Additional Description:
Risk Level: Low/Medium
Testing:
Docs Changes: NA
Release Notes:
Platform Specific Features:

### CODE REVIEW

**Comment by nezdolik** [comment]
@antoniovicente @mattklein123 as discussed some time ago in https://github.com/envoyproxy/envoy/pull/15707, broke huge change into smaller parts.

File: envoy/server/resource_monitor.h [hunk 1]
```diff
@@ -18,39 +21,58 @@ struct ResourceUsage {
...
-    /**
-     * Called when the request for updated resource usage fails.
-     * @param error the exception caught when trying to get updated resource usage
-     */
-    virtual void onFailure(const EnvoyException& error) PURE;
-  };
+  virtual void onFailure(const EnvoyException& error) PURE;
+};
+
+class ReactiveResourceUpdateCallbacks {
```
Comment (nezdolik) [inline envoy/server/resource_monitor.h:44]: need to rename this into proactive

**Comment by nezdolik** [comment]
need to update `overload_integration_test` to support proactive resource monitors as well

**Comment by antoniovicente** [comment]
Unfortunately, I currently don't have a lot of cycles for Envoy reviews.  Assigning to Kevin to do a first pass.

File: source/server/overload_manager_impl.cc [hunk 2]
```diff
@@ -40,10 +44,48 @@ class ThreadLocalOverloadStateImpl : public ThreadLocalOverloadState {
...
 
+  bool tryAllocateResource(OverloadProactiveResourceName resource_name,
+                           int64_t increment) override {
+    const auto proactive_resource = proactive_resources_->find(resource_name);
+    if (proactive_resource != proactive_resources_->end()) {
+      if (proactive_resource->second.tryAllocateResource(increment)) {
+        return true;
+      } else {
+        return false;
+      }
```
Comment (KBaichoo) [inline source/server/overload_manager_impl.cc:55]: Could just do `return proactive_resource->second.tryAllocateResource(increment);`

File: source/server/overload_manager_impl.cc [hunk 3]
```diff
@@ -268,19 +310,49 @@ OverloadManagerImpl::OverloadManagerImpl(Event::Dispatcher& dispatcher, Stats::S
...
+      proactive_resources_(
+          std::make_unique<
+              absl::node_hash_map<OverloadProactiveResourceName, ProactiveResource>>()) {
   Configuration::ResourceMonitorFactoryContextImpl context(dispatcher, options, api,
                                                            validation_visitor);
+  // We should hide impl details from users, for them there should be no distinction between
+  // proactive and regular resource monitors in configuration API. But internally we will maintain
+  // two distinct collections of proactive and regular resources. Proactive resources are not
+  // subject to periodic flushes and can be recalculated/updated on demand by invoking
+  // `tryAllocateResource/tryDeallocateResource` via thread local overload state.
```
Comment (KBaichoo) [inline source/server/overload_manager_impl.cc:323]: This comment was very helpful, is there a doc somewhere I can read upon the differences of these some more? Thanks!

**Review by KBaichoo — COMMENTED** [review]
Thanks for working on this. Here's a first pass. 

Is there's a doc on the rest of the chain of PRs that I can read to uunderstand the bigger picture? Thanks

File: source/server/overload_manager_impl.cc (hunk 3 above)
Comment (nezdolik) [reply source/server/overload_manager_impl.cc:323]: @KBaichoo we don't have design doc that describes in detail the latest proposed implementation for proactive checks, there were multiple iterations. I will prep a doc, good idea.

**Comment by nezdolik** [comment]
Clang tidy check fails with error which i don't think is related to this change:
```
clang-tidy check failed, potentially fixed by clang-apply-replacements:
Diagnostics:
- BuildDirectory: /build/tmp/_bazel_envoybuild/b570b5ccd0454dc9af9f65ab1833764d/execroot/envoy
  DiagnosticMessage:
    FileOffset: 313
    FilePath: ./source/common/common/hash.h
    Message: '''xxhash.h'' file not found'
    Replacements: []
  DiagnosticName: clang-diagnostic-error
  Level: Error
  Ranges:
  - {FileOffset: 313, FilePath: ./source/common/common/hash.h, Length: 10}
  ```

**Comment by rojkov** [comment]
The CI failure looks irrelevant indeed. Could you perhaps merge the main branch again and see if it helps?

/wait

**Comment by rojkov** [comment]
This time there is a new problem #19060 with CI.

**Review by KBaichoo — APPROVED** [review]
Thanks for bearing with us @nezdolik. @yanavlasov were you planning to review this as well? This LGTM apart from the nits pointed out.

**Comment by nezdolik** [comment]
Thanks @KBaichoo @yanavlasov, applied review comments.

**Review by KBaichoo — APPROVED** [review]

**Comment by nezdolik** [comment]
@yanavlasov would you be able to take a look?

**Review by yanavlasov — APPROVED** [review]

**Pull request merged**